recipes/omniorb/meta.yaml

```diff
@@ -0,0 +1,49 @@
+{% set version = "4.2.3" %}
+
+package:
+  name: omniorb
+  version: {{ version }}
+
+source:
+#  fn: omniORB-{{ version }}-{{ release }}.tar.gz
+  url: https://sourceforge.net/projects/omniorb/files/omniORB/omniORB-{{ version }}/omniORB-{{ version }}.tar.bz2
+  sha256: 26412ac08ab495ce5a6a8e40961fa20b7c43f623c6c26b616d210ca32f078bca
+  patches:
+    - patch-configure.diff
+
+build:
+  number: 0
+  skip: True  # [win or py!=37]
```

Comment:
This line is preventing building with Python 3.6.

Reply:
Oh! Thanks!

Reply:
I actually inserted that line because I thought that python was only a requirement for building the package, not for running it, so one python version was enough (but maybe, as often, I'm wrong!)

Reply:
In that case `python` should go in the `build` section rather than the `host` section. It could also include a version requirement like:
```yaml
requirements:
  build:
    - python  # Can use any version of python, will probably pick the newest
    - python >=3.7
    - python =3.7
    - python >=3.5,<3.8
```
(only one would be needed, I'm just showing how to constrain the version)

Reply:
The omniidl compiler is written in Python (including the parts which generate C++ stubs and skeletons). It is packaged with omniORB, but omniORBpy adds a new backend to generate Python code.

Reply:
yeah, I tried that but it didn't work on linux. Not sure why, though. Mac seemed to build fine (see b9058b5)

Reply:
I have to leave so I don't have time to investigative further at the moment but [this](https://conda.io/projects/conda-build/en/latest/source/resources/define-metadata.html#requirements-section) is the relevant part of the docs.

In summary, there are three options:
 * If you **only** use Python in the build process, it should go in the `build` requirements and there will be a single build per platform.
 * If you link against Python's shared libraries it needs to be in the `host` and `run` requirements. There will be three builds per platform, one for each of Python 2.7, 3.6 and 3.7.
 * If you need Python at runtime but the used when building doesn't affect the binaries (i.e. a pure Python package with no linking to `libpython`), it can be a [`noarch: python` package](https://conda.io/projects/conda-build/en/latest/source/resources/define-metadata.html#architecture-independent-packages). There will be only one build per platform.

Reply:
I'm quite confused about this... take a look at the osx CI here: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=15758&view=logs from this commit here: https://github.com/conda-forge/staged-recipes/tree/b9058b5d18114ebe4a00a0a4a6e938cb15648d21/recipes/omniorb

@chrisburr is it possible that the osx CI was green because it was picking up a system-installed python?

Reply:
Carlo, where do you put your conda_build_config.yaml? I've stuffed one into the omniorb recipe directory but it seems that is not the right place. Though it might be the right place if we want to iterate on it to get things building.
I've installed the SDK for conda on Mac and added it to the conda_build_config.yaml per instructions. But after re-enabling the compiler lines in meta.yaml I'm back to getting clang-4.0 (somewhere in the conda docs it says "Don't. Just don't." or something to that effect). That fails in omniORB autoconf:
`
configure:3145: x86_64-apple-darwin13.4.0-clang -V >&5
clang-4.0: error: argument to '-V' is missing (expected 1 value)
clang-4.0: error: no input files
configure:3156: $? = 1
configure:3145: x86_64-apple-darwin13.4.0-clang -qversion >&5
clang-4.0: error: unknown argument: '-qversion'
clang-4.0: error: no input files
`

Reply:
as I showed you above, the osx builds are already passing (at least before I started shooting commits in the dark to try to fix Linux builds).

Please tell me: are the tests in the `tests` section of `meta.yaml` enough to judge that OmniORB is working fine?

Reply:
No that test is not sufficient. It does test that the IDL compiler (written in Python) can generate stubs and skeletons for C++, but does not try to link them or to use them.

Also, the configure line is producing broken installations, since PREFIX seems to point at the build area, not the area in which the software is installed. So omniORB.cfg will not be found, omniNames will not be able to write a log file, and quite possibly python will not be properly referenced since it is having an absolute path written into the first line of omniidl.

Can we retrieve a CI services built binary to test on our local machines? If so I can analyze some of the installed artifacts to see what needs to be fixed. Or I could build locally as I've been trying to do which gives me direct and immediate access to the artifacts.

Reply:
1. I've just pushed a commit which fixes build error, the issue you were seeing on Linux is that the kernel truncates shebangs that are longer than 128 characters and conda builds with a 256 character install location. ([Lovely case of the Baader–Meinhof effect.](https://lwn.net/SubscriberLink/779997/11de2bdc8dbc0d69/)) I've added a crude patch that uses `/usr/bin/env python` instead. Feel free to replace it if you'd rather handle it differently.

2. This now builds locally for me. If you're still having trouble, share the errors you get and I'll take a look.

3. For the installation, it's normally fine if the installed artefacts contain absolute paths. When packaging conda automatically modifies the run path in executables/libraries and also detects two cases:
   - **binary mode**: Replace the build's install prefix with the user install prefix, appending null bytes to the end so the size of the file remains the same.
   - **text mode**: Replace the build's install prefix with the user install prefix just removing any excess bytes.
Occasionally you have to be careful of this behaviour, like putting the absolute path in a `std::string` fails when the install prefix path is shorter. There are several ways of working around it if needed.

4. Lastly, from what you're saying it sounds like the compiler should be a runtime dependency?

Reply:
> ... It could also include a version requirement like:
> 
> ```yaml
> requirements:
>   build:
>     - python  # Can use any version of python, will probably pick the newest
>     - python >=3.7
>     - python =3.7
>     - python >=3.5,<3.8
> ```

So a line like `- python =2.7,>=3.5` fails. Removing the "=2.7," seems to work. How best to enable 2.7.x and some of the 3.x series?

Reply:
I'm not sure why I should not expect a local build of a recipe to work. conda supplies a complete build and runtime environment so as long as the dependencies are specified properly it should Just Work.

So, current sticking points on MacOS High Sierra (10.13.6) include getting a bogus compiler clang-4.0 if
```
- {{ compiler('c') }}
- {{ compiler('cxx') }}
```
are specified in meta.xml. If those are omitted, it picks up gcc on my computer, possibly from my MacPorts installation since I don't see it mentioned in the packages being installed by conda-build.

And specifying
```
conda build . --python=3.6
```
gets me a build with Python-3.7.

I re-introduced the conda_build_config.yaml with just the following lines enabled:
```
CONDA_BUILD_SYSROOT:
  - /opt/MacOSX10.9.sdk        # [osx]
channel_sources:
- conda-forge/label/cf201901,defaults   # [unix]
- conda-forge,defaults                  # [win]
channel_targets:
- conda-forge cf201901                  # [unix]
- conda-forge main                      # [win]
python:
  - 3.6
```
and continue to get a Python-3.7 build with a gcc compiler though my build line is
```
conda build . --python=3.6
```

I'm sure these are newbie problems for me, but a complete recipe should mention all of the prerequisites and have a successful build which relies on the conda installation(s) so we aren't there yet. Right?

Reply:
The lines containing `cf201901` are actively harmful and will prevent things from working.

As a sanity check, can update and try building from a fresh clone of the repository using these steps:

```bash
conda activate root
conda update -c conda-forge conda-build conda
conda --version  # should be conda 4.6.7
conda build --version  # should be conda-build 3.17.8

git clone git@github.com:carlodri/staged-recipes.git -b omniorb
# Or this if you don't have SSH set up
# git clone https://github.com/carlodri/staged-recipes.git -b omniorb
cd staged-recipes
conda build -c conda-forge -c defaults recipes/omniorb --python=3.6
```

On macOS the `$CXX` variable is equal to `x86_64-apple-darwin13.4.0-clang++` which is itself a symlink pointing to `clang-4.0`. If you look at [line 1281 in this log on Travis](https://travis-ci.org/conda-forge/staged-recipes/builds/498923922#L1281) you should be able to compare to figure out what is going on.

Reply:
OK, I followed your instructions above to the letter and conda now respects the --python=3.6 line. Yay!

I've also removed the lines from meta.yaml which have explicit "cf201901" specifications. My build process is now explicitly bringing in the c and c++ compilers to the conda build, which it was not doing earlier. Yay! Will commit that change.

I was misled by the autoconf log (there are some tests of the compiler command line parameters which are intended to fail while determining the compiler features). I'm failing the configure test for the c compiler because it is not finding "-lSystem". When I omit the compiler specification in meta.yaml, then it finds and uses gcc from elsewhere, which succeeds.

btw I'm using a conda_build_config.yaml file in the staged-recipes/ directory which *only* defines the MacOS SDK per the conda on-line documentation.

Thanks everyone for the help so far; getting closer but not there yet!